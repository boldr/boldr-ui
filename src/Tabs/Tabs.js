// @flow
import React from 'react';
import type { Node } from 'react';
import RcTabs, { TabPane } from 'rc-tabs';
import ScrollableInkTabBar from 'rc-tabs/lib/ScrollableInkTabBar';
import TabContent from 'rc-tabs/lib/TabContent';
import classNames from 'classnames';
import Icon from '../Icons/Icon';

export type TabsType = 'line' | 'card' | 'editable-card';
export type TabsPosition = 'top' | 'right' | 'bottom' | 'left';

export type TabsProps = {
  activeKey?: string,
  defaultActiveKey?: string,
  hideAdd?: boolean,
  children: Array<React.Element<*>>,
  onChange?: (activeKey: string) => void,
  onTabClick?: Function,
  onPrevClick?: SyntheticEvent<>,
  onNextClick?: SyntheticEvent<>,
  tabBarExtraContent?: Node | null,
  tabBarStyle?: Object,
  type?: TabsType,
  tabPosition?: TabsPosition,
  onEdit?: (targetKey: string, action: any) => void,
  size?: 'default' | 'small',
  style?: Object,
  prefixCls?: string,
  className?: string,
  animated?: boolean | { inkBar: boolean, tabPane: boolean },
};

// Tabs
export type TabPaneProps = {
  tab?: Node | string,
  style?: Object,
  closable?: boolean,
  className?: string,
  disabled?: boolean,
};

export default class Tabs extends React.Component<TabsProps, any> {
  static TabPane = (TabPane: TabPaneProps);

  static defaultProps = {
    prefixCls: 'boldrui-tabs',
    hideAdd: false,
  };

  createNewTab = targetKey => {
    const onEdit = this.props.onEdit;
    if (onEdit) {
      onEdit(targetKey, 'add');
    }
  };

  removeTab = (targetKey, e) => {
    e.stopPropagation();
    if (!targetKey) {
      return;
    }

    const onEdit = this.props.onEdit;
    if (onEdit) {
      onEdit(targetKey, 'remove');
    }
  };

  handleChange = activeKey => {
    const onChange = this.props.onChange;
    if (onChange) {
      onChange(activeKey);
    }
  };

  render() {
    let {
      prefixCls,
      className = '',
      size,
      type = 'line',
      tabPosition,
      children,
      tabBarExtraContent,
      tabBarStyle,
      hideAdd,
      onTabClick,
      onPrevClick,
      onNextClick,
      animated = true,
    } = this.props;

    let { inkBarAnimated, tabPaneAnimated } =
      typeof animated === 'object'
        ? {
            inkBarAnimated: animated.inkBar,
            tabPaneAnimated: animated.tabPane,
          }
        : {
            inkBarAnimated: animated,
            tabPaneAnimated: animated,
          };

    // card tabs should not have animation
    if (type !== 'line') {
      tabPaneAnimated = 'animated' in this.props ? tabPaneAnimated : false;
    }

    let cls = classNames(className, {
      ['boldrui-tabs-mini']: size === 'small' || size === 'mini',
      ['boldrui-tabs-vertical']: tabPosition === 'left' || tabPosition === 'right',
      ['boldrui-tabs-card']: type.indexOf('card') >= 0,
      [`boldrui-tabs-${type}`]: true,
      ['boldrui-tabs-no-animation']: !tabPaneAnimated,
    });
    // only card type tabs can be added and closed
    let childrenWithClose;
    if (type === 'editable-card') {
      childrenWithClose = [];
      React.Children.forEach(children, (child: React.Element<*>, index) => {
        let closable = child.props.closable;
        closable = typeof closable === 'undefined' ? true : closable;
        const closeIcon = closable ? (
          <Icon kind="close" onClick={e => this.removeTab(child.key, e)} />
        ) : null;
        childrenWithClose.push(
          React.cloneElement(child, {
            tab: (
              <div className={closable ? undefined : 'boldrui-tabs-tab-unclosable'}>
                {child.props.tab}
                {closeIcon}
              </div>
            ),
            key: child.key || index,
          }),
        );
      });
      // Add new tab handler
      if (!hideAdd) {
        tabBarExtraContent = (
          <span>
            <Icon kind="plus" className="boldrui-tabs-new-tab" onClick={this.createNewTab} />
            {tabBarExtraContent}
          </span>
        );
      }
    }

    tabBarExtraContent = tabBarExtraContent ? (
      <div className="boldrui-tabs-extra-content">{tabBarExtraContent}</div>
    ) : null;

    const renderTabBar = () => (
      <ScrollableInkTabBar
        inkBarAnimated={inkBarAnimated}
        extraContent={tabBarExtraContent}
        onTabClick={onTabClick}
        onPrevClick={onPrevClick}
        onNextClick={onNextClick}
        style={tabBarStyle}
      />
    );

    return (
      <RcTabs
        {...this.props}
        className={cls}
        tabBarPosition={tabPosition}
        renderTabBar={renderTabBar}
        renderTabContent={() => <TabContent animated={tabPaneAnimated} animatedWithMargin />}
        onChange={this.handleChange}
      >
        {childrenWithClose || children}
      </RcTabs>
    );
  }
}
