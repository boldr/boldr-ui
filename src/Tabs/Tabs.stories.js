import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import centered from '@storybook/addon-centered';
import Tabs from './Tabs';

const TabPane = Tabs.TabPane;

storiesOf('Tabs', module)
  .addDecorator(centered)
  .add('default', () => (
    <Tabs defaultActiveKey="1" onChange={action('onChange')}>
      <TabPane tab="Tab 1" key="1">
        Content of Tab Pane 1
      </TabPane>
      <TabPane tab="Tab 2" key="2">
        Content of Tab Pane 2
      </TabPane>
      <TabPane tab="Tab 3" key="3">
        Content of Tab Pane 3
      </TabPane>
    </Tabs>
  ));
