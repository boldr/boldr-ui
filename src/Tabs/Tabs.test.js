/* eslint-env jest */
import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import Tabs from './Tabs';

const { TabPane } = Tabs;
describe('Tabs', () => {
  describe('tabPosition', () => {
    it('remove card', () => {
      const wrapper = mount(
        <Tabs tabPosition="left" tabBarExtraContent="xxx">
          <TabPane tab="foo" key="1">
            foo
          </TabPane>
        </Tabs>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
