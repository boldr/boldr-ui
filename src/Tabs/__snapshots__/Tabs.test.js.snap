// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs tabPosition remove card 1`] = `
<Tabs
  hideAdd={false}
  prefixCls="boldrui-tabs"
  tabBarExtraContent="xxx"
  tabPosition="left"
>
  <Tabs
    className="boldrui-tabs-vertical boldrui-tabs-line"
    destroyInactiveTabPane={false}
    hideAdd={false}
    onChange={[Function]}
    prefixCls="boldrui-tabs"
    renderTabBar={[Function]}
    renderTabContent={[Function]}
    style={Object {}}
    tabBarExtraContent="xxx"
    tabBarPosition="left"
    tabPosition="left"
  >
    <div
      className="boldrui-tabs boldrui-tabs-left boldrui-tabs-vertical boldrui-tabs-line"
      style={Object {}}
    >
      <ScrollableInkTabBar
        activeKey="1"
        extraContent={
          <div
            className="boldrui-tabs-extra-content"
          >
            xxx
          </div>
        }
        inkBarAnimated={true}
        onKeyDown={[Function]}
        onNextClick={[Function]}
        onPrevClick={[Function]}
        onTabClick={[Function]}
        panels={
          <TabPane
            placeholder={null}
            tab="foo"
          >
            foo
          </TabPane>
        }
        prefixCls="boldrui-tabs"
        scrollAnimated={true}
        styles={Object {}}
        tabBarPosition="left"
      >
        <div
          className="boldrui-tabs-bar"
          onKeyDown={[Function]}
          role="tablist"
          tabIndex="0"
        >
          <div
            className="boldrui-tabs-nav-container"
          >
            <span
              className="boldrui-tabs-tab-prev boldrui-tabs-tab-btn-disabled"
              onClick={null}
              onTransitionEnd={[Function]}
              unselectable="unselectable"
            >
              <span
                className="boldrui-tabs-tab-prev-icon"
              />
            </span>
            <span
              className="boldrui-tabs-tab-next boldrui-tabs-tab-btn-disabled"
              onClick={null}
              unselectable="unselectable"
            >
              <span
                className="boldrui-tabs-tab-next-icon"
              />
            </span>
            <div
              className="boldrui-tabs-nav-wrap"
            >
              <div
                className="boldrui-tabs-nav-scroll"
              >
                <div
                  className="boldrui-tabs-nav boldrui-tabs-nav-animated"
                >
                  <div
                    className="boldrui-tabs-ink-bar boldrui-tabs-ink-bar-animated"
                  />
                  <div
                    aria-disabled="false"
                    aria-selected="true"
                    className="boldrui-tabs-tab-active boldrui-tabs-tab"
                    onClick={[Function]}
                    role="tab"
                  >
                    foo
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="boldrui-tabs-extra-content"
            style={Object {}}
          >
            xxx
          </div>
        </div>
      </ScrollableInkTabBar>
      <TabContent
        activeKey="1"
        animated={true}
        animatedWithMargin={true}
        destroyInactiveTabPane={false}
        onChange={[Function]}
        prefixCls="boldrui-tabs"
        tabBarPosition="left"
      >
        <div
          className="boldrui-tabs-content boldrui-tabs-content-animated"
          style={
            Object {
              "marginTop": "0%",
            }
          }
        >
          <TabPane
            active={true}
            destroyInactiveTabPane={false}
            placeholder={null}
            rootPrefixCls="boldrui-tabs"
            tab="foo"
          >
            <div
              aria-hidden="false"
              className="boldrui-tabs-tabpane boldrui-tabs-tabpane-active"
              role="tabpanel"
            >
              foo
            </div>
          </TabPane>
        </div>
      </TabContent>
    </div>
  </Tabs>
</Tabs>
`;
